### PR TITLE
ESQL: Handle allocation errors inside topn

### PR DESCRIPTION
**Pull request by nik9000**
This properly handles allocation errors inside of topn by making `Block.Builder` and `Vector.Builder` `Releasable`. The "new way" to deal with block factories is like this:
```
try (var b = IntBlock.builder(3, blockFactory) {
  b.append(1);
  b.append(2);
  b.append(3);
  return b.build();
}
```

If anything goes wrong the block factory's `close` method will be called by the `try` block and all of the circuit breaking that it reserves will be released.

For this all to work well `Block.Builder`s have to be one-shot. In other words, you can only call `.build` on them one time. That shifts the accounting from the builder into the block. It is an error to call `build` twice.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/CannedSourceOperator.java [hunk 1]
```diff
@@ -96,5 +96,9 @@ public Page getOutput() {
     }
 
     @Override
-    public void close() {}
+    public void close() {
+        while (page.hasNext()) {
+            page.next().releaseBlocks();
+        }
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/CannedSourceOperator.java:102]: This let's me assert that we've closed all the input pages even if these's an error!

**Comment by nik9000** [comment]
run elasticsearch-ci/part-2

**Comment by nik9000** [comment]
```
» [2023-09-26T21:41:09,555][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [javaRestTest-1] fatal error in thread [elasticsearch[javaRestTest-1][esql_worker][T#24]], exiting java.lang.AssertionError: Used bytes: [-24] must be >= 0
```

There's some kind of double release going on, I think it's on blocks but I'm not sure.

**Comment by nik9000** [comment]
CrankyBreaker is my hero! 
```
    @Repeat(iterations=1000)
    public void testCranky() {
        BigArrays bigArrays = new MockBigArrays(PageCacheRecycler.NON_RECYCLING_INSTANCE, new CrankyCircuitBreakerService());
        BlockFactory blockFactory = new BlockFactory(bigArrays.breakerService().getBreaker(CircuitBreaker.REQUEST), bigArrays);
        try {
            try (Block.Builder builder = elementType.newBlockBuilder(10, blockFactory)) {
                BasicBlockTests.RandomBlock random = BasicBlockTests.randomBlock(elementType, 10, false, 1, 1, 0, 0);
                builder.copyFrom(random.block(), 0, random.block().getPositionCount());
                try (Block built = builder.build()) {
                    assertThat(built, equalTo(random.block()));
                }
            }
            // If we made it this far cranky didn't fail us!
        } catch (CircuitBreakingException e) {
            assertThat(e.getMessage(), equalTo(CrankyCircuitBreakerService.ERROR_MESSAGE));
        }
        assertThat(blockFactory.breaker().getUsed(), equalTo(0L));
    }
```

**Comment by nik9000** [comment]
OK! The problem is with the BytesRefBuilder. We "estimate" the bytes used, but they aren't really an estimate at all.

**Comment by nik9000** [comment]
I got it! Fix incoming.

**Comment by nik9000** [comment]
Almost there! BytesRefs weren't building in the same way as everything else but they sure tried to. It wasn't that big a deal that we were off before, but this should catch it! Lots more tests incoming too.

**Comment by nik9000** [comment]
OK! That should work.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java [hunk 2]
```diff
@@ -399,14 +406,22 @@ private Iterator<Page> toPages() {
...
             return result.iterator();
         } finally {
-            Releasables.closeExpectNoException(() -> Releasables.close(list));
+            if (success == false) {
+                List<Releasable> close = new ArrayList<>(list);
+                for (Page p : result) {
+                    close.add(p::releaseBlocks);
+                }
+                Collections.addAll(close, builders);
+                Releasables.closeExpectNoException(Releasables.wrap(close));
```
Comment (ChrisHegarty) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java:423]: This turns out to be not so bad. 👍

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BlockFactoryTests.java [hunk 3]
```diff
@@ -37,6 +37,10 @@
 // BlockFactory is used and effectively tested in many other places, but this class contains tests
 // more specific to the factory implementation itself (and not necessarily tested elsewhere).
 public class BlockFactoryTests extends ESTestCase {
+    public static BlockFactory blockFactory(ByteSizeValue size) {
+        BigArrays bigArrays = new MockBigArrays(PageCacheRecycler.NON_RECYCLING_INSTANCE, size);
+        return new BlockFactory(bigArrays.breakerService().getBreaker(CircuitBreaker.REQUEST), bigArrays);
+    }
```
[the remark this answers is not in the document]
Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BlockFactoryTests.java:43]: ++

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BytesRefBlockEqualityTests.java [hunk 4]
```diff
@@ -332,10 +331,14 @@ public void testSimpleBlockWithSingleNull() {
     public void testSimpleBlockWithManyNulls() {
         int positions = randomIntBetween(1, 256);
         boolean grow = randomBoolean();
-        var builder = BytesRefBlock.newBlockBuilder(grow ? 0 : positions);
-        IntStream.range(0, positions).forEach(i -> builder.appendNull());
-        BytesRefBlock block1 = builder.build();
-        BytesRefBlock block2 = builder.build();
+        BytesRefBlock.Builder builder1 = BytesRefBlock.newBlockBuilder(grow ? 0 : positions);
+        BytesRefBlock.Builder builder2 = BytesRefBlock.newBlockBuilder(grow ? 0 : positions);
```
[the remark this answers is not in the document]
Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BytesRefBlockEqualityTests.java:335]: This is actually more readably. 👍

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java [hunk 5]
```diff
@@ -88,21 +88,21 @@ public final void testSimpleCircuitBreaking() {
...
+        BlockFactory blockFactory = BlockFactory.getInstance(breaker, bigArrays);
+        Exception e = expectThrows(
+            CircuitBreakingException.class,
+            () -> drive(simple(bigArrays).get(new DriverContext(bigArrays, blockFactory)), input.iterator())
+        );
+        assertThat(e.getMessage(), equalTo(MockBigArrays.ERROR_MESSAGE));
+
+        // Note the lack of try/finally here - we're asserting that when the driver throws an exception we clear the breakers.
+        assertThat(bigArrays.breakerService().getBreaker(CircuitBreaker.REQUEST).getUsed(), equalTo(0L));
+        assertThat(inputFactoryContext.breaker().getUsed(), equalTo(0L));
```
Comment (ChrisHegarty) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:105]: ++

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/topn/TopNOperatorTests.java [hunk 6]
```diff
@@ -1299,13 +1349,23 @@ public void testCloseWithoutCompleting() {
...
-        return new DriverContext(bigArrays, new BlockFactory(bigArrays.breakerService().getBreaker(CircuitBreaker.REQUEST), bigArrays));
+        CircuitBreaker breaker = bigArrays.breakerService().getBreaker(CircuitBreaker.REQUEST);
+        breakers.add(breaker);
+        return new DriverContext(bigArrays, new BlockFactory(breaker, bigArrays));
+    }
+
+    @After
+    public void allBreakersEmpty() {
+        for (CircuitBreaker breaker : breakers) {
+            assertThat(breaker.getUsed(), equalTo(0L));
```
Comment (ChrisHegarty) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/topn/TopNOperatorTests.java:1367]: This is going to be useful. 👍

**Review by ChrisHegarty — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BlockFactoryTests.java [hunk 7]
```diff
@@ -48,11 +48,29 @@ public static BlockFactory blockFactory(ByteSizeValue size) {
...
+            @Override
+            public BlockFactory get() {
+                CircuitBreaker breaker = new MockBigArrays.LimitedBreaker("esql-test-breaker", ByteSizeValue.ofGb(1));
+                BigArrays bigArrays = new MockBigArrays(PageCacheRecycler.NON_RECYCLING_INSTANCE, mockBreakerService(breaker));
+                return BlockFactory.getInstance(breaker, bigArrays);
+            }
+
+            @Override
+            public String toString() {
+                return "1gb";
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BlockFactoryTests.java:61]: Yeah! I got the old error message from gradle and thought "I have no idea what this means". Easy enough fix!

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/topn/TopNOperatorTests.java (hunk 6 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/topn/TopNOperatorTests.java:1367]: Yeah. I think we should drag this into `AnyOperatorTests` - maybe pretty soon.

**Comment by nik9000** [comment]
The part two errors look real! I'll take a look.

**Comment by nik9000** [comment]
run elasticsearch-ci/part-3

**Pull request merged**